main.cpp

```diff
@@ -0,0 +1,327 @@
+
+Tank tanks[NUMBER_OF_TANKS] = {
+    {AnalogIn(A0), DigitalOut(D2), DigitalOut(LED1), WATER_PUMP_BUTTON, 0, 0, 0, 0, 0, 0},
+    {AnalogIn(A1), DigitalOut(D3), DigitalOut(LED2), SOLENOID_VALVE_1_BUTTON, 0, 0, 0, 0, 0, 0},
+    {AnalogIn(A2), DigitalOut(D4), DigitalOut(LED3), SOLENOID_VALVE_2_BUTTON, 0, 0, 0, 0, 0, 0},
+}; /**< Array of tanks. Helps avoiding code repetition and managing the level sensors readings and actuators activation.*/
+
+UnbufferedSerial uartUsb(USBTX, USBRX, 115200);
+
+bool manualMode = false;
+
+void inputsInit();
+void outputsInit();
+void getTanksWaterLevel();
+void manageWaterLevel();
+void availableCommands();
+void debounceButton(Button*);
+void buttonsInit();
+void uartTask();
+void manualLevelManagement();
+void checkManualManagementSwitch();
+
+int main()
+{
+    inputsInit();
+    outputsInit();
+    buttonsInit();
+    while(true) {
+        if(manualMode == true) {
+            manualLevelManagement();
+        } else {
+            checkManualManagementSwitch();
+            getTanksWaterLevel();
+            manageWaterLevel();
+        }
+        uartTask();
+    }
+}
+
+void inputsInit() {
+}
+
+/**
+* @brief Iterates over the tanks array and initiate the actuators and leds with the OFF value.
+*/
+void outputsInit() {
+    for (int i = 0; i < NUMBER_OF_TANKS; i++) {
+        tanks[i].actuator = OFF;
+        tanks[i].led = OFF;
+    }
+}
+
+/**
+* @brief Function used to calculate the water level in the tanks, saving the information in the structs of the tanks array.
+*
+* The sensors readings are saved into the levelSensorsReadingsArray. All the readings in the array are then averaged to
+* obtain a more reliable analog measure.
+*
+* The arrays are filled replacing the older measures by the new ones. At the begining, when the array is empty, the measure
+* is not accurate. In our case, the system gets stable once the array is filled (imperceptible time).
+*
+* The Tank struct and the tanks array helps to avoid code repetition here. All the measures are saved in the Tank structs
+* initialized in the tanks array.
+*/
+void getTanksWaterLevel() {
+    static int levelSensorsSampleIndex = 0;
+
+    for (int i = 0; i < NUMBER_OF_TANKS; i++){
+        tanks[i].levelSensorReadingsArray[levelSensorsSampleIndex] = tanks[i].levelSensor.read();
+    }
+
+    levelSensorsSampleIndex += 1;
+
+    if (levelSensorsSampleIndex >= NUMBER_OF_AVG_SAMPLES) {
+        levelSensorsSampleIndex = 0;
+    }
+
+    for(int i = 0; i < NUMBER_OF_TANKS; i++) {
+        tanks[i].levelSensorReadingsSum = 0;
+    }
+
+    for(int i = 0; i < NUMBER_OF_TANKS; i++){
+        for (int j = 0; j < NUMBER_OF_AVG_SAMPLES; j++) {
+            tanks[i].levelSensorReadingsSum += tanks[i].levelSensorReadingsArray[j];
+        }
+    }
+    
+    for(int i = 0; i < NUMBER_OF_TANKS; i++) {
+        tanks[i].levelSensorReadingsAverage = tanks[i].levelSensorReadingsSum / NUMBER_OF_AVG_SAMPLES;
+    }
+}
+
```

Comment:
You start accumulating on an empty array. You perform averaging on every call.
When the array is empty, you average to a lower number, what is wrong. It takes NUMBER_OF_AVG_SAMPLES to get the actual value, ramping up from 0.
Once the array has been filled, you average samples at different "windows".
Why the average ? Why 100 samples ?
What is your sampling rate ? It depends on whether something is asked over the UART, and the length of its response...